// @jsmini/is/dist/index.aio.js
/*!
 * is 0.4.0 (https://github.com/jsmini/is)
 * API https://github.com/jsmini/is/blob/master/doc/api.md
 * Copyright 2017-2018 yanhaijing. All Rights Reserved
 * Licensed under MIT (https://github.com/jsmini/is/blob/master/LICENSE)
 */

(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(factory((global.jsmini_is = {})));
}(this, (function (exports) { 'use strict';

/*!
 * type 0.4.0 (https://github.com/jsmini/type)
 * API https://github.com/jsmini/type/blob/master/doc/api.md
 * Copyright 2017-2018 jsmini. All Rights Reserved
 * Licensed under MIT (https://github.com/jsmini/type/blob/master/LICENSE)
 */

const toString = Object.prototype.toString;

function type(x) {
    if(x === null){
        return 'null';
    }

    const t= typeof x;

    if(t !== 'object'){
        return t;
    }

    let c;
    try {
        c = toString.call(x).slice(8, -1).toLowerCase();
    } catch(exp) {
        return 'object';
    }

    if(c !== 'object'){
        return c;
    }

    if(x.constructor == Object){
        return c;
    }

    return 'unknown';
}

function isNumber(x) {
    return type(x) === 'number';
}

function isBoolean(x) {
    return type(x) === 'boolean';
}

function isString(x) {
    return type(x) === 'string';
}

function isNull(x) {
    return type(x) === 'null';
}

function isUndefined(x) {
    return type(x) === 'undefined';
}

function isObject(x) {
    return type(x) === 'object';
}

function isFunction(x) {
    return type(x) === 'function';
}

const isArray = isFunction(Array.isArray) ? Array.isArray : function isArray(x) {
    return type(x) === 'array';
};

exports.isNumber = isNumber;
exports.isBoolean = isBoolean;
exports.isString = isString;
exports.isNull = isNull;
exports.isUndefined = isUndefined;
exports.isObject = isObject;
exports.isFunction = isFunction;
exports.isArray = isArray;

})));
