// @jsmini/is/dist/index.js
/*!
 * is 0.4.0 (https://github.com/jsmini/is)
 * API https://github.com/jsmini/is/blob/master/doc/api.md
 * Copyright 2017-2018 yanhaijing. All Rights Reserved
 * Licensed under MIT (https://github.com/jsmini/is/blob/master/LICENSE)
 */

'use strict';

var type = require('@jsmini/type');

function isNumber(x) {
    return type.type(x) === 'number';
}

function isBoolean(x) {
    return type.type(x) === 'boolean';
}

function isString(x) {
    return type.type(x) === 'string';
}

function isNull(x) {
    return type.type(x) === 'null';
}

function isUndefined(x) {
    return type.type(x) === 'undefined';
}

function isObject(x) {
    return type.type(x) === 'object';
}

function isFunction(x) {
    return type.type(x) === 'function';
}

const isArray = isFunction(Array.isArray) ? Array.isArray : function isArray(x) {
    return type.type(x) === 'array';
};

exports.isNumber = isNumber;
exports.isBoolean = isBoolean;
exports.isString = isString;
exports.isNull = isNull;
exports.isUndefined = isUndefined;
exports.isObject = isObject;
exports.isFunction = isFunction;
exports.isArray = isArray;
