// @jsmini/inherits/dist/index.aio.js
/*!
 * inherits 0.5.0 (https://github.com/jsmini/inherits)
 * API https://github.com/jsmini/inherits/blob/master/doc/api.md
 * Copyright 2017-2018 yanhaijing. All Rights Reserved
 * Licensed under MIT (https://github.com/jsmini/inherits/blob/master/LICENSE)
 */

(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(factory((global.jsmini_inherits = {})));
}(this, (function (exports) { 'use strict';

/*!
 * type 0.4.0 (https://github.com/jsmini/type)
 * API https://github.com/jsmini/type/blob/master/doc/api.md
 * Copyright 2017-2018 jsmini. All Rights Reserved
 * Licensed under MIT (https://github.com/jsmini/type/blob/master/LICENSE)
 */

const toString = Object.prototype.toString;

function type(x) {
    if(x === null){
        return 'null';
    }

    const t= typeof x;

    if(t !== 'object'){
        return t;
    }

    let c;
    try {
        c = toString.call(x).slice(8, -1).toLowerCase();
    } catch(exp) {
        return 'object';
    }

    if(c !== 'object'){
        return c;
    }

    if(x.constructor == Object){
        return c;
    }

    return 'unknown';
}

/*!
 * is 0.4.0 (https://github.com/jsmini/is)
 * API https://github.com/jsmini/is/blob/master/doc/api.md
 * Copyright 2017-2018 yanhaijing. All Rights Reserved
 * Licensed under MIT (https://github.com/jsmini/is/blob/master/LICENSE)
 */

function isFunction(x) {
    return type(x) === 'function';
}

const isArray = isFunction(Array.isArray) ? Array.isArray : function isArray(x) {
    return type(x) === 'array';
};

function create(prototype) {
    //临时构造函数
    var F = function F() {};
    F.prototype = prototype;
    return new F();
}

function setPrototypeOf(C, P, force) {
    // 静态属性继承，慎用，有坑
    if (Object.setPrototypeOf) {
        // setPrototypeOf es6
        Object.setPrototypeOf(C, P);
    } else if (C.__proto__) {
        // __proto__ es6引入，但是部分浏览器早已支持
        C.__proto__ = P;
    } else if (force) {
        // 兼容ie10-等陈旧浏览器
        // 将P上的静态属性和方法拷贝一份到C上，不会覆盖C上的方法
        for (var k in P) {
            if (P.hasOwnProperty(k) && !(k in C)) {
                C[k] = P[k];
            }
        }
    }
}

function inherits(C, P) {
    // 如果不为函数，则不进行操作
    if (!isFunction(C)) {
        throw new TypeError('inherits first param must is function');
    }

    if (!isFunction(P)) {
        throw new TypeError('inherits second param must is function');
    }

    // 等同于临时构造函数
    C.prototype = Object.create(P.prototype);

    C.prototype.constructor = C; // 修复constructor
    C.uper = P; //存储超类

    // 静态属性继承，慎用，有坑，无法兼容10-
    setPrototypeOf(C, P);
}

function inherits$1(C, P) {
    // 如果不为函数，则不进行操作
    if (!isFunction(C)) {
        throw new TypeError('inherits first param must is function');
    }

    if (!isFunction(P)) {
        throw new TypeError('inherits second param must is function');
    }

    C.prototype = create(P.prototype);

    C.prototype.constructor = C; //修复constructor
    C.uper = P; //存储超类

    // 静态属性继承，慎用，有坑，无法兼容10-
    setPrototypeOf(C, P);
}

function inherits$2(C, P) {
    // 如果不为函数，则不进行操作
    if (!isFunction(C)) {
        throw new TypeError('inherits first param must is function');
    }

    if (!isFunction(P)) {
        throw new TypeError('inherits second param must is function');
    }

    C.prototype = create(P.prototype);

    C.prototype.constructor = C; //修复constructor
    C.uper = P; //存储超类

    // 静态属性继承，慎用，有坑
    setPrototypeOf(C, P, true);
}

exports.inherits = inherits;
exports.inheritsByES3 = inherits$1;
exports.inheritsByForce = inherits$2;

})));
